elab_expr.cc

```diff
@@ -5958,6 +5958,23 @@ NetExpr* PEIdent::elaborate_expr_net_idx_up_(Design*des, NetScope*scope,
 	    if (base_c->value().is_defined()) {
 		  long lsv = base_c->value().as_long();
 		  long rel_base = 0;
+
+		    // Check whether an unsigned base fits in a 32 bit int,
+		    // which is what will be used for addressing later on.
```

Comment:
This seems to be unnecessary. The compiler already handles your test case for a constant base correctly. If you compile with -Wall, it will issue an appropriate warning too. 

The same goes for the similar changes below.

If you disagree, please provide a test case that fails.

Reply:
I finally found the time to look at this again with a fresh pair of eyes. I have addressed the comments in your thorough review, and I have cleaned up, corrected, and commented the code to make it more understandable.

As mentioned earlier, if you run the test for constant base through `-t vlog95` first, it will fail. I believe it will also fail right away on Windows (LLP64), however I can't test that.

Reply:
It doesn't fail on Windows because `verinum::as_long()` takes care not to change the sign when it discards excess bits. Both before and after applying this PR, it prints
````
warning: verinum::as_long() truncated 32 bits to 31, returns 2147483647
````
and the test for constant base passes.

(as an aside, it would probably be better if `verinum::as_long()` saturated rather than truncated, but that's for another day)

It's failing when run through `-t vlog95` because the vlog95 code generator is casting the `unsigned` value returned by `ivl_lpm_base()` to an `int`. That looks like a bug in the code generator to me.

I'm not convinced this is the correct solution (for this particular aspect - the rest of the PR is OK), but I guess we could accept it for now and fix it properly later.
@caryr , what do you think?

Reply:
@martinwhitaker Would it help if I simply removed those checks, to get the important part of the PR merged?

Reply:
No, I'll accept this for now, and revisit it when I have more time.

Reply:
OK, thanks.